expression/expression.go

```diff
@@ -208,6 +209,125 @@ func EvalBool(ctx sessionctx.Context, exprList CNFExprs, row chunk.Row) (bool, b
 	return true, false, nil
 }
 
+var (
+	selPool = sync.Pool{
+		New: func() interface{} {
+			return make([]int, 1024)
+		},
+	}
+)
+
+// VecEvalBool does the same thing as EvalBool but it works in a vectorized manner.
+func VecEvalBool(ctx sessionctx.Context, exprList CNFExprs, input *chunk.Chunk, selected, nulls []bool) ([]bool, []bool, error) {
+	n := input.NumRows()
+	selected = selected[:0]
+	nulls = nulls[:0]
+	for i := 0; i < n; i++ {
+		selected = append(selected, false)
+		nulls = append(nulls, false)
+	}
+
+	defer input.SetSel(input.Sel())
+	sel := selPool.Get().([]int)
+	defer selPool.Put(sel)
+	sel = sel[:0]
+	for i := 0; i < n; i++ {
+		sel = append(sel, i)
+	}
+	input.SetSel(sel)
+
+	for _, expr := range exprList {
+		eType := expr.GetType().EvalType()
+		buf, err := globalColumnAllocator.get(eType, n)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if err := vecEval(ctx, expr, input, buf); err != nil {
+			return nil, nil, err
+		}
+
+		isEQCondFromIn := IsEQCondFromIn(expr)
+		hasUnsignedFlag := mysql.HasUnsignedFlag(expr.GetType().Flag)
+		d, j := types.Datum{}, 0
+		for i := range sel {
+			if buf.IsNull(i) {
+				d.SetValue(nil)
+			} else {
+				switch eType {
+				case types.ETInt:
+					if hasUnsignedFlag {
+						d.SetUint64(buf.GetUint64(i))
+					} else {
+						d.SetInt64(buf.GetInt64(i))
+					}
+				case types.ETReal:
+					d.SetFloat64(buf.GetFloat64(i))
+				case types.ETDuration:
+					d.SetMysqlDuration(buf.GetDuration(i, 0))
+				case types.ETDatetime, types.ETTimestamp:
+					d.SetMysqlTime(buf.GetTime(i))
+				case types.ETString:
+					d.SetString(buf.GetString(i))
+				case types.ETJson:
+					d.SetMysqlJSON(buf.GetJSON(i))
+				case types.ETDecimal:
+					d.SetMysqlDecimal(buf.GetDecimal(i))
+				}
+			}
+
+			if d.IsNull() {
+				if !isEQCondFromIn { // same as EvalBool
```

Comment:
how about move this code into `if buf.IsNull(i) {`
or use `SetNull` replace `SetValue(nil)`

Reply:
updated